lib/utmp.c

```diff
@@ -161,21 +161,25 @@ get_current_utmp(pid_t main_pid)
 		    && (USER_PROCESS  != ut->ut_type))
 			continue;
 
-		if (   (main_pid == ut->ut_pid)
-		    && ('\0' != ut->ut_id[0])
-		    /* A process may have failed to close an entry
-		     * Check if this entry refers to the current tty */
-		    && is_my_tty(ut->ut_line))
-		{
-			break;
+		if (main_pid == ut->ut_pid) {
+			if (is_my_tty(ut->ut_line))
+				break; /* Perfect match, stop the search */
+
+			if (NULL == ut_by_pid)
+				ut_by_pid = ut;
+
+		} else if (   (NULL == ut_by_line)
+			   && (LOGIN_PROCESS == ut->ut_type) /* Be more picky when matching by 'ut_line' only */
+			   && (is_my_tty(ut->ut_line))) {
+			ut_by_line = ut;
 		}
 	}
 
-	if (NULL != ut) {
+	if (ut || ut_by_pid || ut_by_line) {
 		struct utmpx  *ut_copy;
 
 		ut_copy = XMALLOC(1, struct utmpx);
-		memcpy(ut_copy, ut, sizeof(*ut));
+		memcpy(ut_copy, ut ?: ut_by_pid ?: ut_by_line, sizeof(*ut));
```

Comment:
I preferred your previous approach.  Now we have two points where we have split the logic about ut?:ut_by_pid?:ut_by_line.